python/lsst/analysis/tools/tasks/sourceObjectTableAnalysis.py

```diff
@@ -20,19 +20,127 @@
+__all__ = (
+    "SourceObjectTableAnalysisConfig",
+    "SourceObjectTableAnalysisTask",
+    "ObjectEpochTableConfig",
+    "ObjectEpochTableTask",
+)
 
 import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
 import numpy as np
 import pandas as pd
 from astropy.table import vstack
-from lsst.pipe.base import NoWorkFound
+from lsst.drp.tasks.gbdesAstrometricFit import calculate_apparent_motion
 from lsst.pipe.base import connectionTypes as ct
 from smatch import Matcher
 
 from ..interfaces import AnalysisBaseConfig, AnalysisBaseConnections, AnalysisPipelineTask
 
 
+class ObjectEpochTableConnections(
+    pipeBase.PipelineTaskConnections,
+    dimensions=("tract", "skymap"),
+    defaultTemplates={
+        # "inputName": "sourceTable_visit",
+        # "inputCoaddName": "deep",
+        # "associatedSourcesInputName": "isolated_star_sources",
+        # "outputName": "sourceObjectTable",
+    },
+):
+    refCat = ct.Input(
+        doc="Catalog of positions to use as reference.",
+        name="objectTable",
+        storageClass="DataFrame",
+        dimensions=["skymap", "tract", "patch"],
+        multiple=True,
+        deferLoad=True,
+        deferGraphConstraint=True,
+    )
+
+    refCatEpochMap = ct.Input(
+        doc="Healsparse map of mean epoch of refCat.",
+        name="deepCoadd_epoch_map_mean",
+        storageClass="HealSparseMap",
+        dimensions=("skymap", "tract", "band"),
+        multiple=True,
+        deferLoad=True,
+    )
+
+    objectEpochs = ct.Output(
+        doc="Catalog of epochs for refCat objects.",
+        name="objectEpochTable",
+        storageClass="DataFrame",
+        dimensions=["skymap", "tract", "patch"],
+        multiple=True,
+    )
+
+
+class ObjectEpochTableConfig(pipeBase.PipelineTaskConfig, pipelineConnections=ObjectEpochTableConnections):
+    refCatBands = pexConfig.ListField(
+        doc=("Bands in refCat to be combined with `refCat_selectors` to build refCat column names."),
+        dtype=str,
+        default=["u", "g", "r", "i", "z", "y"],
+    )
+
+
+class ObjectEpochTableTask(pipeBase.PipelineTask):
+    """Collect mean epochs for the observations that went into each object."""
+
+    ConfigClass = ObjectEpochTableConfig
+    _DefaultName = "objectEpochTable"
+
+    def getEpochs(self, cat, epochMapDict):
+        """Get mean epoch of the visits corresponding to object position.
+
+        Parameters
+        ----------
+        cat : `pd.DataFrame`
+            Catalog containing object positions.
+        refCatEpochMap: `dict` [`DeferredDatasetHandle`]
+            Dictionary of handles for healsparse maps containing the mean epoch
+            for positions in the reference catalog.
+
+        Returns
+        -------
+        epochDf = `pd.DataFrame`
+            Catalog with mean epoch of visits at each object positions.
+        """
+        epochs = {}
+        for band in self.config.refCatBands:
+            refEpochs = np.ones(len(cat)) * np.nan
```

Comment:
If `validPositions.any()` is False, you wouldn't go into the block where that happens. I wanted to do
`refEpochs[validPositions][refEpochsValid] = bandEpochs`
to avoid having a second line setting `bandEpochs[~refEpochsValid] = np.nan`, which I do currently, but this doesn't work.